FWCore/Services/plugins/StallMonitor.cc

```diff
@@ -0,0 +1,403 @@
+// -*- C++ -*-
+//
+// Package: FWCore/Services
+// Class  : StallMonitor
+//
+// Implementation:
+//
+// Original Author:  Kyle Knoepfel
+//
+
+#include "DataFormats/Provenance/interface/ModuleDescription.h"
+#include "FWCore/Concurrency/interface/ThreadSafeOutputFileStream.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/ServiceRegistry/interface/ServiceMaker.h"
+#include "FWCore/ServiceRegistry/interface/StreamContext.h"
+#include "FWCore/ServiceRegistry/interface/ModuleCallingContext.h"
+#include "FWCore/ServiceRegistry/interface/SystemBounds.h"
+#include "FWCore/Utilities/interface/Algorithms.h"
+#include "FWCore/Utilities/interface/Column.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "tbb/concurrent_unordered_map.h"
+
+#include <atomic>
+#include <chrono>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace {
+
+  using clock_t = std::chrono::steady_clock;
+  auto now = clock_t::now;
+
+  inline auto stream_id(edm::StreamContext const& cs)
+  {
+    return cs.streamID().value();
+  }
+
+  inline auto module_id(edm::ModuleCallingContext const& mcc)
+  {
+    return mcc.moduleDescription()->id();
+  }
+
+  //===============================================================
+  class StallStatistics {
+  public:
+
+    // c'tor receiving 'std::string const&' type not provided since we
+    // must be able to call (e.g.) std::vector<StallStatistics>(20),
+    // for which a default label is not sensible in this context.
+
+    std::string const& label() const { return label_; }
+    unsigned numberOfStalls() const { return stallCounter_; }
+
+    using duration_t = std::chrono::milliseconds;
+    using rep_t = duration_t::rep;
+
+    duration_t totalStalledTime() const { return duration_t{totalTime_.load()}; }
+    duration_t maxStalledTime() const { return duration_t{maxTime_.load()}; }
+
+    // Modifiers
+    void setLabel(std::string const& label) { label_ = label; }
+
+    void update(std::chrono::milliseconds const ms)
+    {
+      ++stallCounter_;
+      auto const thisTime = ms.count();
+      totalTime_ += thisTime;
+      rep_t max {maxTime_};
+      while (thisTime > max && !maxTime_.compare_exchange_weak(max, thisTime)) {}
```

Comment:
I'm not a big fan of `compare_exchange_weak` especially on Intel hardware where I believe such an instruction doesn't really exist. I suggest always sticking to `compare_exchange_strong` since the semantics are easier to get right.